### DISCUSSION TITLE
ImageLoader does not return exact pixel values?

### DISCUSSION THREAD

**Discussion by nilscb** [visgl/loaders.gl#2067] [Q&A] [answered]
I have been struggling a bit trying to read off pixel values from a png file using the ImageLoader. But for some reason the pixels values comes out a bit off. I know what the values in the png are. Lets say a png pixel has this value 50,61,71,77 for R,G,B,A respectively, The ImageLoader will resolve the image into a ImageData structure with a "data" member of type Uint8ClampedArray which should contain the raw pixel data. But the pixel comes as 50, 60, 70, 77. In other words slightly off. It does not help no matter which options I set on the ImageLoader. Anyone has a clue to why this is so. Is it not possible to use the ImageLoader to get exact pixel values from a png (or any other image format)?

regards 
Nils Christian Bonnevie
frustrated programmer.

**Comment by ibgreen** [comment, marked as answer]
Thanks for reporting, wasn't aware of this. For image loading, loaders.gl relies on the browser's built in decoding. Perhaps the browser takes some shortcuts with rounding etc to parse faster?

IIRC on a modern desktop browser loaders.gl parses images to ImageBitmap before calling getImageData and the ImageBitmap constructor has a few [options](https://developer.mozilla.org/en-US/docs/Web/API/createImageBitmap) that you should be able to pass through with `options.image.imageBitmap: {...}`. But I can't see anything that would reasonably cause such off-by-one errors...

**Reply by nilscb** [reply]
I did try the ImageBitmap options and they do change the result somewhat. But still not the correct one. I guess that the process the ImageLoader (and many other it seems) uses is effective but not quite exact. I am decoding a float 32 bit in the pixel so I will need exact values. But maybe I can look into parsing and reading the png file directly myself. Thanks a lot for your answer now I know I will have to go to plan B and that is helpful :-)

**Reply by ibgreen** [reply]
When running under Node.js, browser functions to parse images are not available. In this case, loaders.gl uses https://github.com/arian/pngjs to parse PNGs 

(assuming `@loaders.gl/polyfills` has been installed - which installs `get-pixels` which installs `pngjs`).

**Reply by nilscb** [reply]
This is spot on what I need in this particular case. Thanks again!
(I also found this https://github.com/vivaxy/png but any of them will do)
